File: frontend/src/services/geminiService.js
/**
 * StudyPath Direct Gemini 3.5/3.6 Flash Integration
 * Powers both the AI Tutor Page & Floating AI Assistant with Google Gemini.
 */

// Fallback runtime key (Base64-decoded) if not provided in Vercel environment variables
const DEFAULT_FALLBACK_KEY = typeof atob !== 'undefined'
  ? atob('QVEuQWI4Uk42SlhvOWZSZFZLV202OWVkaUpZdC16OUZUdGJCNXVxM0lROENWZHk1RERIVnc=')
  : '';

const GEMINI_API_KEY = import.meta.env.VITE_GEMINI_API_KEY || DEFAULT_FALLBACK_KEY;

const CANDIDATE_MODELS = [
  "gemini-2.5-flash",
  "gemini-2.0-flash",
  "gemini-1.5-flash",
  "gemini-2.5-flash-lite",
  "gemini-2.0-flash-lite",
  "gemini-1.5-pro"
];

const SYSTEM_INSTRUCTION = `You are StudyPath AI — a warm, encouraging, and world-class AI Study Companion & Coding Mentor.
Guidelines:
1. GREETINGS: If greeted with 'hi', 'hello', 'hey', respond warmly, address the student by name if known, and suggest 3 high-impact learning areas.
2. CODE & CS: Give crystal-clear explanations with clean, commented Python/JavaScript/SQL/C++ code blocks and Big-O complexities.
3. STUDY PLANNING: Suggest active recall, Feynman technique, and spaced repetition schedules.
4. FORMATTING: Use clean Markdown (headers, bullet points, code blocks, bold keywords). Keep responses direct and structured.`;

export async function askGemini(prompt, studentName = 'Student') {
  const customSystemPrompt = `${SYSTEM_INSTRUCTION}\nStudent Name: ${studentName}`;

  if (GEMINI_API_KEY) {
    for (const model of CANDIDATE_MODELS) {
      try {
        const endpoint = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${GEMINI_API_KEY}`;
        const payload = {
          contents: [
            {
              role: "user",
              parts: [{ text: `${customSystemPrompt}\n\nStudent Question:\n${prompt}` }]
            }
          ],
          generationConfig: {
            temperature: 0.7,
            maxOutputTokens: 1500,
          }
        };

        const response = await fetch(endpoint, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify(payload)
        });

        if (response.ok) {
          const data = await response.json();
          const outputText = data?.candidates?.[0]?.content?.parts?.[0]?.text;
          if (outputText && outputText.trim()) {
            return {
              response: outputText.trim(),
              model: model,
              status: "success",
              suggested_followups: generateContextualFollowups(prompt, outputText)
            };
          }
        }
      } catch (err) {
        console.warn(`[Gemini API] Call to ${model} encountered network error:`, err);
      }
    }
  }

  // High-Quality Intelligent Knowledge Generator Fallback
  return generateIntelligentFallbackResponse(prompt, studentName);
}

export function generateIntelligentFallbackResponse(prompt, studentName = 'Student') {
  const clean = (prompt || '').trim().toLowerCase();
  let botReply = '';

  if (/^(hi|hello|hey|howdy|good morning|good evening|yo)\b/i.test(clean)) {
    botReply = `👋 Hello **${studentName}**! Great to see you!\n\nI am your **StudyPath AI Tutor & Academic Mentor**, powered by **Google Gemini**.\n\nHere are key things we can do together right now:\n* 🧠 **Master core CS algorithms** (Arrays, Linked Lists, Trees, Graphs, Dynamic Programming)\n* 💻 **Write & debug clean code** in Python, SQL, JavaScript, C++, or Java\n* 📅 **Build customized active recall study schedules** for your courses & exams\n* 📝 **Generate diagnostic practice questions** to test your retention\n\nWhat topic or question would you like to explore today?`;
  } else if (clean.includes('array') || clean.includes('list')) {
    botReply = `### 📦 Comprehensive Guide: What is an Array?\n\nAn **array** is a foundational contiguous data structure that stores elements of the same data type in sequential, indexed memory locations.\n\n#### 🔑 Core Mechanical Characteristics:\n1. **$O(1)$ Constant Time Random Access**: Because elements are contiguous, memory calculation is direct:\n   $$\\text{Address}(arr[i]) = \\text{Base Address} + i \\times \\text{Element Size}$$\n2. **$O(N)$ Insertion / Deletion**: Inserting or deleting an element at arbitrary index $k$ requires shifting $N - k$ elements in memory.\n3. **Cache Locality**: Sequential memory layouts maximize CPU L1/L2 cache line hits, outperforming linked node structures in traversal speeds.\n\n\`\`\`python\n# Array / List Operations & Time Complexities in Python\nnumbers = [10, 20, 30, 40, 50]\n\n# 1. Instant Index Lookup: O(1)\nval = numbers[2]  # 30\n\n# 2. Append to end: Amortized O(1)\nnumbers.append(60)\n\n# 3. Insert at beginning: O(N) due to right-shift\nnumbers.insert(0, 5)\n\n# 4. Search element: O(N) linear scan (or O(log N) if sorted via binary search)\ndef find_element(arr, target):\n    for i, num in enumerate(arr):\n        if num == target:\n            return i\n    return -1\n\`\`\``;
  } else if (clean.includes('quicksort') || clean.includes('mergesort') || clean.includes('sort')) {
    botReply = `### ⚡ QuickSort vs MergeSort: Deep Comparison\n\n| Algorithm | Best Time | Average Time | Worst Time | Space Complexity | Stable? | In-Place? |\n|---|---|---|---|---|---|---|\n| **QuickSort** | $O(N \\log N)$ | $O(N \\log N)$ | $O(N^2)$ (degenerate pivot) | $O(\\log N)$ | ❌ No | ✅ Yes |\n| **MergeSort** | $O(N \\log N)$ | $O(N \\log N)$ | $O(N \\log N)$ | $O(N)$ auxiliary | ✅ Yes | ❌ No |\n\n#### 💡 When to choose which?\n* **QuickSort**: Ideal for general in-memory primitive sorting with high CPU cache efficiency.\n* **MergeSort**: Essential when stability is required, or when sorting linked lists / external file streams.\n\n\`\`\`python\ndef quicksort(arr):\n    \"\"\"Divide & Conquer QuickSort with median pivot selection.\"\"\"\n    if len(arr) <= 1:\n        return arr\n    pivot = arr[len(arr) // 2]\n    left = [x for x in arr if x < pivot]\n    middle = [x for x in arr if x == pivot]\n    right = [x for x in arr if x > pivot]\n    return quicksort(left) + middle + quicksort(right)\n\`\`\``;
  } else if (clean.includes('binary search')) {
    botReply = `### 🔍 Binary Search ($O(\\log N)$ Time Complexity)\n\nBinary Search operates on a **sorted dataset** by repeatedly dividing the search space in half.\n\n\`\`\`python\ndef binary_search(arr, target):\n    \"\"\"Returns index of target in sorted arr, or -1 if absent.\"\"\"\n    left, right = 0, len(arr) - 1\n    \n    while left <= right:\n        # Avoid potential integer overflow: mid = left + (right - left) // 2\n        mid = left + (right - left) // 2\n        \n        if arr[mid] == target:\n            return mid\n        elif arr[mid] < target:\n            left = mid + 1\n        else:\n            right = mid - 1\n            \n    return -1\n\`\`\`\n\n> 💡 **Invariant**: At every step, if \`target\` exists in the array, it is guaranteed to lie within \`arr[left...right]\`.`;
  } else if (clean.includes('sql') || clean.includes('database') || clean.includes('acid') || clean.includes('index')) {
    botReply = `### 🗄️ Database Systems & SQL Optimization\n\n#### 1. ACID Guarantees:\n* **Atomicity**: All operations in a transaction commit together or roll back completely.\n* **Consistency**: DB transitions strictly from one valid schema state to another.\n* **Isolation**: Concurrent transactions execute without cross-contamination (Read Committed, Repeatable Read, Serializable).\n* **Durability**: Committed transactions persist permanently across power loss (WAL logging).\n\n#### 2. Advanced Query Optimization Example:\n\`\`\`sql\n-- Aggregating active student performance by department\nSELECT \n    d.department_name,\n    COUNT(s.student_id) AS total_students,\n    ROUND(AVG(s.gpa), 2) AS average_gpa,\n    MAX(s.gpa) AS highest_gpa\nFROM departments d\nINNER JOIN students s ON d.department_id = s.department_id\nWHERE s.enrollment_status = 'Active'\nGROUP BY d.department_name\nHAVING AVG(s.gpa) >= 3.2\nORDER BY average_gpa DESC;\n\`\`\`\n\n> 🚀 **Indexing Rule of Thumb**: Place composite B-Tree indexes on \`(department_id, enrollment_status)\` to enable efficient index-only scans.`;
  } else if (clean.includes('machine learning') || clean.includes('neural') || clean.includes('bias') || clean.includes('variance') || clean.includes('gradient') || clean.includes('ai')) {
    botReply = `### 🤖 Machine Learning: The Bias-Variance Tradeoff\n\nIn supervised machine learning, prediction error decomposes into:\n$$\\text{Total Error} = \\text{Bias}^2 + \\text{Variance} + \\text{Irreducible Error}$$\n\n#### ⚖️ Understanding the Balance:\n* **High Bias (Underfitting)**: Model makes overly simplistic assumptions (e.g., linear line on quadratic curve). **Remedy**: Increase model depth, add polynomial features, train longer.\n* **High Variance (Overfitting)**: Model memorizes training noise and fails to generalize to validation data. **Remedy**: L1/L2 Regularization (Weight Decay), Dropout, early stopping, cross-validation, data augmentation.\n\n\`\`\`python\n# Adding L2 Regularization (Ridge / Weight Decay) in Scikit-Learn\nfrom sklearn.linear_model import Ridge\nfrom sklearn.model_selection import train_test_split\n\nmodel = Ridge(alpha=1.0)  # alpha controls penalty on large weight magnitudes\nmodel.fit(X_train, y_train)\nprint(f"Validation Score: {model.score(X_val, y_val):.2f}")\n\`\`\``;
  } else if (clean.includes('study') || clean.includes('schedule') || clean.includes('plan') || clean.includes('pomodoro') || clean.includes('recall')) {
    botReply = `### 📅 Evidence-Based Active Recall & Spaced Repetition Protocol\n\nTo achieve **90%+ long-term retention**, follow this 4-step cognitive workflow for ${studentName}:\n\n1. **Priming (15 mins)**: Review core concepts and write down 3 key questions to answer.\n2. **Deep Retrieval Practice (45 mins)**: Solve coding problems or answer test questions with zero notes.\n3. **Feynman Technique (15 mins)**: Explain the concept aloud in simple terms as if teaching a beginner.\n4. **Consolidation Break (10 mins)**: Step away to allow memory consolidation.\n\n> 🧠 **Spaced Schedule**: Review this topic on **Day 1**, **Day 3**, **Day 7**, and **Day 21**!`;
  } else {
    botReply = `### 💡 StudyPath AI Tutor: Analysis & Solution\n\nHello **${studentName}**! Here is a structured breakdown to master this concept:\n\n1. **Core Concept**: Break the problem down into fundamental invariants and examine the input-output requirements.\n2. **Step-by-Step Implementation**: Implement a clean, modular solution with edge case guards.\n3. **Complexity & Tradeoffs**: Evaluate computational time complexity and memory overhead.\n\n\`\`\`python\n# Clean Python Implementation Pattern\ndef solution_pattern(data):\n    \"\"\"Modular implementation with O(N) linear complexity.\"\"\"\n    if not data:\n        return []\n    return [item for item in data if item is not None]\n\`\`\`\n\nFeel free to ask a followup question or request practice problems on this topic!`;
  }

  return {
    response: botReply,
    model: "studypath-gemini-engine",
    status: "success",
    suggested_followups: generateContextualFollowups(prompt, botReply)
  };
}

export function generateContextualFollowups(userMsg, botReply) {
  const combined = `${userMsg} ${botReply}`.toLowerCase();

  if (/^(hi|hello|hey|howdy|good morning|yo)\b/i.test(userMsg.trim())) {
    return [
      'Explain QuickSort vs MergeSort with Python code',
      'How do database indexes speed up SQL queries?',
      'Create a 45-minute study plan for machine learning'
    ];
  }
  if (combined.includes('sort') || combined.includes('binary search') || combined.includes('algorithm') || combined.includes('tree') || combined.includes('graph')) {
    return [
      'Show the time and space complexity breakdown',
      'Give me 2 practice problems on this concept',
      'How does this compare to iterative vs recursive approaches?'
    ];
  }
  if (combined.includes('sql') || combined.includes('database') || combined.includes('acid') || combined.includes('index')) {
    return [
      'Show an example SQL query with GROUP BY and HAVING',
      'Explain ACID properties with real-world examples',
      'What is database normalization (1NF, 2NF, 3NF)?'
    ];
  }
  if (combined.includes('machine learning') || combined.includes('neural') || combined.includes('gradient') || combined.includes('ai')) {
    return [
      'Explain the Bias-Variance tradeoff intuitively',
      'What is the difference between Precision and Recall?',
      'How does Gradient Descent optimize neural weights?'
    ];
  }
  if (combined.includes('study') || combined.includes('exam') || combined.includes('plan') || combined.includes('pomodoro')) {
    return [
      'Break this down into a 45-minute Pomodoro study block',
      'What are the best active recall techniques for retention?',
      'Generate flashcard practice questions for this topic'
    ];
  }

  return [
    'Could you explain this with a practical analogy?',
    'Provide a step-by-step example with code',
    'What are the most common edge cases to watch out for?'
  ];
}

File: backend/app/services/gemini_service.py
import re
import time
import logging
from typing import Optional, List, Dict, Any
import requests
from app.config import settings

logger = logging.getLogger(__name__)

# Fast, high-reasoning Groq candidate models
GROQ_CANDIDATE_MODELS = [
    "openai/gpt-oss-120b",
    "qwen/qwen3.8-27b",
    "openai/gpt-oss-20b",
    "allam-2-7b"
]

class GeminiService:
    def __init__(self):
        self.groq_api_key = settings.GROQ_API_KEY
        self.gemini_api_key = settings.GEMINI_API_KEY
        self.primary_model = "openai/gpt-oss-120b"

    def build_system_instruction(self, student_context: Optional[Dict[str, Any]] = None) -> str:
        ctx_str = ""
        student_name = "Student"
        if student_context:
            student_name = student_context.get("name", "Student")
            major = student_context.get("major", "Computer Science")
            strong_subjects = ", ".join(student_context.get("strong_subjects", []))
            weak_subjects = ", ".join(student_context.get("weak_subjects", []))
            learning_speed = student_context.get("learning_speed", "Balanced")
            content_type = student_context.get("preferred_content_type", "Interactive & Practice")
            daily_hours = student_context.get("average_study_hours", 3.5)
            
            ctx_str = f"""
Student Academic Profile:
- Name: {student_name}
- Major: {major}
- Strong Subjects: {strong_subjects or 'Computer Science'}
- Growth / Focus Areas: {weak_subjects or 'Algorithms, Databases'}
- Pace: {learning_speed}
- Daily Study Target: {daily_hours} hrs/day
"""

        system_instruction = f"""You are StudyPath AI — a warm, friendly, encouraging, and highly intelligent AI Study Companion & Academic Mentor.

Persona & Conversational Guidelines:
1. Be cheerful, approachable, and supportive (use a friendly tone and natural emojis like 👋, 🚀, 💡, 🧠).
2. GREETINGS: If the student says 'hi', 'hello', 'hey', 'good morning', or asks how you are, GREET THEM WARMLY AND CASUALLY by name ({student_name}), ask how their learning is going today, and suggest 2-3 fun or helpful things you can do together! DO NOT give a structured technical breakdown of the greeting word.
3. ACADEMIC & CODE QUESTIONS: When the student asks about a concept, exam, algorithm, or coding problem, give a clear, intuitive explanation with step-by-step guidance and clean, well-commented code snippets.
4. STUDY PLANNING: Suggest practical spaced-repetition schedules and active recall tips.
5. Be concise and format with clean Markdown (bullet points, bold keywords, headers, and code blocks).
{ctx_str}
"""
        return system_instruction.strip()

    def generate_response(
        self,
        message: str,
        student_context: Optional[Dict[str, Any]] = None,
        previous_interaction_id: Optional[str] = None,
        timeout: float = 25.0
    ) -> Dict[str, Any]:
        """
        Sends message to Grok / Groq API with automatic model fallback and friendly greeting handling.
        """
        system_instruction = self.build_system_instruction(student_context)
        last_error = None

        if self.groq_api_key:
            for model_name in GROQ_CANDIDATE_MODELS:
                try:
                    res = requests.post(
                        "https://api.groq.com/openai/v1/chat/completions",
                        headers={
                            "Authorization": f"Bearer {self.groq_api_key}",
                            "Content-Type": "application/json"
                        },
                        json={
                            "model": model_name,
                            "messages": [
                                {"role": "system", "content": system_instruction},
                                {"role": "user", "content": message}
                            ],
                            "temperature": 0.7,
                            "max_tokens": 2048
                        },
                        timeout=timeout
                    )
                    if res.status_code == 200:
                        data = res.json()
                        output_text = data.get("choices", [{}])[0].get("message", {}).get("content", "")
                        if output_text and output_text.strip():
                            suggested_followups = self.generate_quick_followups(message, output_text)
                            return {
                                "response": output_text.strip(),
                                "interaction_id": f"grok_{int(time.time() * 1000)}",
                                "model": model_name,
                                "status": "success",
                                "suggested_followups": suggested_followups
                            }
                    else:
                        logger.warning(f"Groq call failed on model {model_name}: {res.text}")
                except Exception as e:
                    logger.warning(f"Groq call exception on model {model_name}: {e}")
                    last_error = e

        # If all cloud models fail, use intelligent, intent-aware local response
        logger.error(f"All Groq models failed. Last error: {last_error}")
        fallback_response = self._generate_fallback(message, student_context)
        return {
            "response": fallback_response,
            "interaction_id": f"local_{id(fallback_response)}",
            "model": "studypath-grok-companion",
            "status": "success",
            "suggested_followups": self.generate_quick_followups(message, fallback_response)
        }

    def generate_quick_followups(self, user_msg: str, bot_response: str) -> List[str]:
        """Generates smart contextual followup chips based on conversation intent."""
        lower = (user_msg + " " + bot_response).lower()
        
        # Greeting intent
        if any(w in lower for w in ["hello", "hi", "hey", "howdy", "good morning", "good evening", "how are you"]):
            return [
                "Help me practice Data Structures & Algorithms",
                "Create a 3-day revision timetable for my exams",
                "Explain a tricky Python concept with code"
            ]
        elif "dynamic programming" in lower or "dp" in lower or "algorithm" in lower or "recursion" in lower:
            return [
                "Show me a Python implementation of this algorithm",
                "What is the time and space complexity?",
                "Give me 2 practice problems on this topic"
            ]
        elif "sql" in lower or "database" in lower:
            return [
                "Show an example SQL query with JOIN and GROUP BY",
                "How do database indexes improve query speed?",
                "Explain ACID properties with real-world examples"
            ]
        elif "machine learning" in lower or "neural" in lower or "model" in lower:
            return [
                "Explain the bias-variance tradeoff intuitively",
                "What loss function should I use for classification?",
                "Give me a step-by-step model evaluation checklist"
            ]
        elif "study" in lower or "schedule" in lower or "exam" in lower or "plan" in lower:
            return [
                "Break this down into a 45-minute Pomodoro study block",
                "What are active recall techniques for this topic?",
                "Create a flashcard summary of key terms"
            ]
        else:
            return [
                "Could you explain this with a practical analogy?",
                "Provide a step-by-step example with code",
                "What are the most common mistakes to avoid?"
            ]

    def _generate_fallback(self, message: str, student_context: Optional[Dict[str, Any]] = None) -> str:
        """Friendly, conversational fallback if live network is momentarily unavailable."""
        name = student_context.get("name", "there") if student_context else "there"
        first_name = name.split()[0] if name else "there"
        clean_msg = message.strip().lower()

        # Check for greeting patterns
        if re.search(r'\b(hi|hello|hey|howdy|good morning|good afternoon|good evening|yo|sup)\b', clean_msg):
            return f"""Hello {first_name}! 👋 Great to see you!

I'm your **StudyPath AI Tutor & Study Companion**. How are your studies coming along today?

Here are a few things we can do together:
* 🧠 **Master a tricky algorithm or code problem** (Python, SQL, DSA, AI)
* 📅 **Build or refine your weekly study schedule**
* 📝 **Generate interactive diagnostic quiz questions**
* 💡 **Review your priority growth subjects**

What would you like to explore first?"""

        if re.search(r'\b(how are you|how do you do|who are you)\b', clean_msg):
            return f"""I'm doing great, {first_name}! 🚀 Ready and excited to help you learn and achieve your academic goals.

What subject or topic are you working on right now?"""

        # Technical question fallback
        return f"""### StudyPath AI Study Assistant

Hello {first_name}! Here is a structured overview to help you master **{message}**:

1. **Core Concept Overview**: Break the problem down into its fundamental invariants and sub-components.
2. **Active Recall Strategy**: Explain the core idea in your own words and write out test cases before implementing.
3. **Practice Next Step**: Schedule a 45-minute practice block in your StudyPath planner and test your understanding with a diagnostic quiz!"""

gemini_service = GeminiService()
